boot/flags.go

```diff
@@ -0,0 +1,251 @@
+		// the factory boot flag is set to indicate that this is a
+		// boot inside a trusted factory environment
+		"factory",
+	}
+)
+
+// splitFlagString splits the given comma delimited list of boot flags, removing
+// empty strings.
+// Note that this explicitly does not filter out unsupported boot flags in the
+// off chance that an old version of the initramfs is reading new boot flags
+// written by a new version of snapd in userspace on a previous boot.
+func splitFlagString(s string) []string {
+	flags := []string{}
+	for _, flag := range strings.Split(s, ",") {
+		if flag != "" {
+			flags = append(flags, flag)
+		}
+	}
+
+	return flags
+}
+
+func checkFlagList(flags []string, allowList []string) error {
+	if len(allowList) != 0 {
+		// then we need to enforce the allow list
+		for _, flag := range flags {
+			if !strutil.ListContains(allowList, flag) {
+				return fmt.Errorf("flag %q is not allowed", flag)
+			}
+		}
+	}
+	return nil
+}
+
+func dropEmptyBootFlags(flags []string) []string {
+	// drop empty strings before serializing
+	nonEmptyFlags := make([]string, 0, len(flags))
+	for _, flag := range flags {
+		if strings.TrimSpace(flag) != "" {
+			nonEmptyFlags = append(nonEmptyFlags, flag)
+		}
+	}
+	return nonEmptyFlags
+}
+
+func serializeFlagString(flags []string) string {
+	return strings.Join(flags, ",")
+}
+
+// SetImageBootFlags writes the provided flags to the bootenv of the recovery
+// bootloader in the specified system rootDir. It is only meant to be called at
+// prepare-image customization time by ubuntu-image/prepare-image.
+func SetImageBootFlags(flags []string, rootDir string) error {
+	// check that the flagList is supported
+	if err := checkFlagList(flags, understoodBootFlags); err != nil {
+		return err
+	}
+
+	// also ensure that the serialized value of the boot flags fits inside the
+	// bootenv value, on lk systems the max size of a bootenv value is 255 chars
+	s := serializeFlagString(dropEmptyBootFlags(flags))
+	if len(s) > 254 {
+		return fmt.Errorf("internal error: boot flags too large to fit inside bootenv value")
+	}
+
+	// now find the recovery bootloader in the system dir and set the value on
+	// it
+	opts := &bootloader.Options{
+		Role: bootloader.RoleRecovery,
+	}
+	bl, err := bootloader.Find(rootDir, opts)
+	if err != nil {
+		return err
+	}
+
+	return bl.SetBootVars(map[string]string{
+		"snapd_boot_flags": s,
+	})
+}
+
+// InitramfsActiveBootFlags returns the set of boot flags that are currently set
+// for the current boot, by querying them directly from the source. This method
+// is only meant to be used from the initramfs, since it may query the bootenv
+// or query the modeenv depending on the current mode of the system.
+// For detecting the current set of boot flags outside of the initramfs, use
+// BootFlags(), which will query for the runtime version of the flags in /run
+// that the initramfs will have setup for userspace.
+// Only to be used on UC20+ systems with recovery systems.
```

Comment:
this should probably mention that no filtering is done on the flags (except dropping empty flags)

Reply:
done